Add collection transformers.

almet opened the pull request:
This adds the concept of collection transformers.

I haven't updated the documentation because I'm not sure about the concepts we want to expose, as mentioned in #154 & #153.

The documentation states that transformers should be able to encode and decode, but I'm not sure that should be the case for what's implemented here as "collection transformers". I'm happy to have any feedback on the naming (and of course the implementation).

If we chose to go with the "transformers" approach (and still not have decode / encode), then the documentation will need to be updated to explain what a collection should be.

For what it's worth, the first version of this code was implemented with "decode" and "encode" functions, but I couldn't see why we were needing them, so I removed them and passed a function instead.

I'm very interested by any feedback on this, so if anyone has time, don't hesitate :) cc @n1k0  @michielbdejong

almet commented:
@n1k0 can you have a look at the code and tell me if that looks okay? (docs needs an update). Thanks!

n1k0 commented on docs/api.md:
```diff
@@ -545,7 +545,7 @@ kinto.events.on("deprecated", function(event) {
 Transformers are basically hooks for encoding and decoding records, which can work synchronously or asynchronously.
 
-For now, only *remote transformers* have been implemented, but there are plans to implement local transformers and collection transformers in a next version.
+For now, only *remote transformers* and *collection transformers* have been implemented, but there are plans to implement local transformers in a next version.
```
Nit: we should reuse the term *hooks*

almet replied:
yes, as I said, this doc is completely out of date. I plan to use hooks here.

leplatrem replied:
Also, I think it could be worth mentioning this feature in the «extending kinto.js» docs, no?

n1k0 commented on src/collection.js:
```diff
@@ -230,6 +237,52 @@ export default class Collection {
+      return {};
+    }
+    if (Array.isArray(hooks)) {
+      throw new Error("hooks should be an object, not an array.");
```
How could this ever happen as this is a private function?

almet replied:
this function is called at initialization, with the data entered by the user in collection.hooks.

n1k0 replied:
I see

almet replied:
I've updated the code to act like you say, but it means that when registering the collection hooks I need to now register a "factory" to access the collection. Something like this:

```js
function signatureVerifierFactory(collection) {
  return function validateCollectionSignature(payload) {
    // Access collection here.
  };
};
```

That's doable, but I wonder if it's the API we want to expose?

n1k0 replied:
Sorry, I'm missing previous context it seems. If you want to pass the collection instance as a second arg to avoid creating factories, that's a very good reason for doing so :)

almet replied:
Okay, going back in the other direction!

n1k0 commented on src/collection.js:
```diff
@@ -537,6 +590,7 @@ export default class Collection {
         if (!syncResultObject.ok) {
           return syncResultObject;
         }
+        // XXX import of data should be done after conflict handling.
```
I don't get this comment

almet replied:
Currently, we import the data *first* and then we check if we have conflicts. I would have expected to retrieve everything, check for conflicts (and solve them) and finally import.

n1k0 replied:
That makes sense, but is probably broader than the scope of this patch; could you please file an issue about it?

n1k0 commented on src/collection.js:
```diff
@@ -633,12 +687,22 @@ export default class Collection {
+      .then(changes => this.applyHook("incoming-changes", {changes}))
       // Reflect these changes locally
-      .then(changes => this.importChanges(syncResultObject, changes))
+      .then(payload => this.importChanges(syncResultObject, payload.changes))
```
Nit: can be written that way

```js
.then(({changes}) => this.importChanges(syncResultObject, changes))
```

n1k0 commented on src/collection.js:
```diff
@@ -633,12 +687,22 @@ export default class Collection {
+      return Promise.resolve(payload);
+    }
+    return waterfall(this.hooks[hookName].map(hook => {
+      return record => hook(payload, this);
```
[not finished reading the whole patch, parking this thought for now]: could we avoid passing `this` as a `hook` dependency here?

almet replied:
it's done on purpose: the hook actually need a reference to the collection to work properly (with my use case in mind, where we need to retrieve the list of records in the database)

n1k0 commented on test/collection_test.js:
```diff
@@ -900,6 +941,42 @@ describe("Collection", () => {
+          articles = testCollection({
+            hooks: {
+              "incoming-changes": [
+                function(payload, collection) {
```
Ok, so I suggest we don't pass collection as an argument, because most often the user will have an instance at hand already.

almet replied:
Yeah, that could also work without passing it it seems.

n1k0 commented on test/collection_test.js:
```diff
@@ -900,6 +941,42 @@ describe("Collection", () => {
+          return articles.pullChanges(result)
+            .then(_ => {
+              expect(hookCalled).to.be.true;
+            });
```
Nit: can be written on a single line

n1k0 commented on test/collection_test.js:
```diff
@@ -900,6 +941,42 @@ describe("Collection", () => {
+            });
+        });
+
+        it("should reject the promise if the hook raises", () => {
```
nit: s/raises/throws ;)

n1k0 commented:
Looks good so far. As hooks seem to support returning promises (because you use `waterfall`), we should probably add some tests for it.

leplatrem commented on docs/api.md:
```diff
@@ -598,7 +598,8 @@ coll.create({title: "foo"}).then(_ => coll.sync())
 
-In a near future, Kinto.js will provide transfomers aimed at providing facilities to encode and decode records when persisted locally.
+In a near future, Kinto.js will provide transfomers aimed at providing facilities to encode and decode records when persisted locally. If you are interested into
+making this happen, [don't hesitate to connect with us](https://kiwiirc.com/client/irc.freenode.net/?#kinto).
```
nit: I would rather point to contributing page instead

n1k0 replied:
+1

almet replied:
Then I need to update the contributing page to have a "how to contact us" link :)

almet commented:
rebased.

almet commented:
@n1k0 last r?

n1k0 commented on docs/extending.md:
````diff
@@ -46,6 +46,7 @@ Then create the Kinto object passing a reference to your adapter class:
 const kinto = new Kinto({adapter: MyAdapter});
 ```
 
+<<<<<<< bb86b694d73101136b4c6c39b47bcf8e27afd4f0
````
?

almet replied:
Ah, this looks like a trace of a rebase :)

n1k0 commented on docs/extending.md:
```diff
@@ -85,3 +86,24 @@ db.list()
+    "hooks": {
+      "incoming-changes": doSomething
+    }
+});
```
Nit: 2 spaces indentation

n1k0 commented on src/collection.js:
```diff
@@ -310,6 +315,52 @@ export default class Collection {
+    if (!Array.isArray(hook)) {
+      throw new Error("A hook definition should be an array of functions.");
+    }
+    return hook.map(fun => {
```
Ubernit: `fn` is more common in js land, and in this project.

n1k0 commented on test/collection_test.js:
```diff
@@ -157,6 +159,45 @@ describe("Collection", () => {
+      function registerHooks(hooks) {
+        return new Collection(TEST_BUCKET_NAME, TEST_COLLECTION_NAME, api, {
+          hooks,
+          adapter: IDB
```
Nit: Isn't this the default?

almet replied:
This is how it's done in the tests at other locations, I just copied it over. Could be addressed as a separate issue I guess.

n1k0 replied:
There are existing tests for async transformers [here](https://github.com/Kinto/kinto.js/blob/master/test/collection_test.js#L930-L944).

almet replied:
It actually breaks with `No adapter provided`.

n1k0 replied:
I suspect that's because we're never expected to use Collection directly, only through the Kinto instance... Ok leave it that way, no big deal.

n1k0 commented on test/collection_test.js:
```diff
@@ -995,6 +1036,40 @@ describe("Collection", () => {
+          });
+
+          return articles.pullChanges(result)
+            .then(_ => {expect(hookCalled).to.be.true;});
```
Nit: unnecessary curly braces. Also, always prefer `eql(true)` for boolean asserts.

n1k0 commented:
I couldn't find tests for using hooks returning a Promise. Did I miss something?

almet commented:
I've updated the pull request:
- [x] addressed your nits;
- [x] added new tests ensuring that the hooks are chainable;
- [x] added new tests ensuring that the return value of the hooks is used;
- [x] pass the current collection as the second argument of the hooks + tests.

Let me know if there is anything else needed in order to land this.

n1k0 commented:
r+
